BasicSample/Assets/Interaction/Scripts/ApplyHaptics.cs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.XR;
+
+namespace Microsoft.MixedReality.OpenXR.BasicSample
+{
+    /// <summary>
+    /// Checks for haptics support in the controller and if supported, it sends an impulse when the trigger is pressed.
+    /// </summary>
+    public class ApplyHaptics : MonoBehaviour
+    {
+        private readonly List<InputDevice> controllerHapticDevices = new List<InputDevice>();
+
+        private void InitializeDevices()
+        {
+            controllerHapticDevices.Clear();
+            List<InputDevice> controllerInputDevices = new List<InputDevice>();
+
+            // Get all controller input devices
+            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.Controller, controllerInputDevices);
+
+            // Filter devices with haptics capabilities
+            foreach (InputDevice controller in controllerInputDevices)
+            {
+                if (controller.TryGetHapticCapabilities(out HapticCapabilities hapticCapabilities) && hapticCapabilities.supportsImpulse)
+                {
+                    controllerHapticDevices.Add(controller);
+                }
+            }
+        }
+
+        private void OnDeviceConnected(InputDevice device)
+        {
+            InitializeDevices();
```

Comment:
With this new event pattern, you shouldn't need to query `InputDevices.GetDevicesWithCharacteristics`. The event itself hands you the new `InputDevice` directly! You should just be able to do

```c#
if (device.TryGetHapticCapabilities(out HapticCapabilities hapticCapabilities) && hapticCapabilities.supportsImpulse)
{
    controllerHapticDevices.Add(device);
}
```

in this callback, and directly `controllerHapticDevices.Remove(device);` in `OnDeviceDisconnected`.

Reply:
Done